### ISSUE TITLE
Engine needs preemptive cancellation

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit-console#642]
I'm marking this as an idea for discussion, although I have pretty much convinced myself it's a necessary feature.

It's a problem I came upon in the GUI. Issue TestCentric/testcentric-gui#337 complained that the GUI was not stopping a test run immediately like NUnit V2 but only after any already running tests complete. The GUI uses a cooperative cancel rather than a forced cancel, so I decided to use force for version 1.0 to be compatible with the V2 GUI.

It turns out that force doesn't work. The framework stops the tests, but it appears not too. The GUI doesn't get the notifications it expects and therefore doesn't think the test has actually stopped. Of course, that's a framework issue, which I'll deal with separately.

Why this issue then?

Thinking about it, I realized that both options, cooperative and preemptive cancellation, are in some way cooperative. The true cooperative cancel requires the cooperation of the tests. The "forced" cancel requires the cooperation of the framework code.

If we consider that runner, engine and framework are independent layers and that drivers can be written for any framework in the future, it seems to me that we need a true preemptive cancel that doesn't depend on the framework at all.

If the test is running in a separate process, that's pretty easy. We could just kill the process or simply tell it to terminate. Telling it seems OK, because the engine is running in both processes.

Things get a bit harder if we are running in process. We might have to use some pretty low-level stuff to get to the executing threads and kill them. Or we might be able to do something at the level of the AppDomain.

As I said at first, this is an idea. I'd like to find out first if people agree with it, then we can worry about how it would be done.

**Comment by ChrisMaddock** [comment]
Does this need an API change? It sounds like something of an implementation detail to me.

I, the user, request the engine to force cancel. The engine passes that to the driver. In the particular implementation of the NUnit 3 driver, the driver chooses to achieve that by passing the command on to the framework. However, there's no reason the driver couldn't do this cancel by killing processes, as you say.

If a framework does not implement it's own cancel (or implements a buggy one), in my mind, that should be for the driver to make up for. I'm not sure the user needs to be exposed to whether the cancellation is called in the framework of directly in the engine.

**Comment by CharliePoole** [comment]
At the user level, I don't think an API change is needed. Either you want to current tests to finish or not. But some internal interfaces might change and I thought that no longer trusting the framework  to take care of it was a big enough change to ask about. Some team members don't always agree on what's an implementation detail. :smile:

**Comment by CharliePoole** [comment]
The changes may go beyond the driver too. For example, cancelling a process belongs in the process runner and will require a method added to ITestAgency.

**Comment by siprbaum** [comment]
To my understanding, when the currently running test has completed, than the TearDown and the 
OneTimeTearDown will be executed. It is pretty common that these methods contain cleanup actions
or free used used resources. If you simply kill the process or forcefully end the test run, these cleanup
actions will not run, which often causes problems when running the tests again.

AFAIK ReSharper uses a different model in the past (can't check right now if it is still the case):

1. Stopping a test run will complete the already running test
2. Stopping again when we already triggered the stopping, but e.g. the currently running tests takes too long to complete will forcefully end the test by killing the process

From a users point, this is actually what I find most appealing. I want a way to end the current
test run, but still have my cleanups executed. In case the test hangs or takes to long to complete,
I still can forcefully end it.

**Comment by CharliePoole** [comment]
Both the NUnit 3 framework and engine support cooperative cancellation (letting the tests complete first) and forced or preemptive cancellation. The code that calls for the run to be stopped chooses which will be used. OTOH NUnit V2 only supported forced cancellation, because that is the only way to stop a test that's stuck in a loop, for example.

In the GUI, I'm aiming to support an approach very much like what you are describing: cooperative first but then forced cancellation if the user wishes. Of course, in order to do that, the engine has to work for both modes. Right now, it doesn't work in the forced mode for NUnit 3 tests, so that can't be used.

This issue discusses ways to provide that support. The original design assumes that the framework contains support for forced cancellation which is currently broken. That's the subject of a separate issue, nunit/nunit#3276. What I'm suggesting here is that the engine should possibly not rely on support in any framework (remember, our engine is intended to work with various frameworks) but should terminate tests forcibly itself - when requested, of course.

Short version of all this: we can't support the two-stage cancellation approach unless the engine is able to do it both ways. :-)

BTW, I don't believe the engine itself should use such a two-stage approach. Things are cleaner if such policy decisions are left to the individual runners and the engine just does what it's told!

**Comment by rprouse** [comment]
I'm not sure that the engine should do more than pass the intention on to the driver through the `void StopRun(bool force)` interface. As you say, some frameworks won't support forced stops, but I think that only the driver has enough knowledge of the framework that it is running to properly stop a running test. The engine cannot know where and how tests are running. In-process, we cannot know which threads have been spun up by the tests and AppDomains may not be involved. If tests are running in another process or even machine, how would we know where or how to kill them?

Or are you thinking of just killing and cleaning up the driver as best you can?

**Comment by CharliePoole** [comment]
I do agree that we should leave it up to the engine. Within the engine, we could leave it entirely up to the driver. Then, when running in a separate process, the driver running in that process would have to take care of it terminating all the threads and then the process could end normally. OTOH, summarily killing the process would do the job rather simply too. Of course, for running in process it's more difficult.

My perspective from the GUI point of view right now (version 1.0) is that I want to duplicate V2 behavior, which simply kills everything.

**Comment by CharliePoole** [comment]
Returning to this now that PR nunit/nunit#3281 fixes the framework problem. That means that `StopRun(true)` will work for versions of the framework that include the fix and fail for those that don't. Of course, "those that don't" includes all current released versions of the framework. :smile: 

So, as far as the engine goes, the problem still exists for all extant versions of the framework and will continue to exist for any users running those releases of the framework in the future.

Following @rprouse's suggestion, here's what I propose...

1. Modify the `NUnit3FrameworkDriver` to recognize whether the framework version it is using has the fix or not.

2. If it does not, put pretty similar code in the driver to what I already added to the parallel dispatcher. That is,
  2.1 Collect info on test suites that start using the `start-suite` event
  2.2 Collect info on test suites that complete using the `test-suite` event
  2.3 Simulate completion with a `Failure:Cancelled` result for any suites that didn't complete after a forced Stop.
  2.4 Do what we can to force the framework into a state where a new run is possible. This may not be necessary, since there is no special action like this taken in the framework fix, but it needs to be tested anyway.

Thoughts?

**Comment by ChrisMaddock** [comment]
Sounds good to me!

**Comment by CharliePoole** [comment]
I've created a branch and will shortly make a PR for what I said I would do in the comment above.

However, I have very strong second thoughts about this approach so I have put back the discussion and design labels while we decide.

The problem is that the issue title (my title, actually) says one thing and the "fix" does something different.

I think that's because there are actually two potential problems to deal with when a framework+driver is told to stop the run...

1. It doesn't actually stop. This is the problem for which I wrote the issue. How does the framework protect itself in this situation? At this point, I believe it's a hypothetical problem, but one we should think about.

2. It stops but the engine doesn't receive proper completion events for all the tests that were started. This is the problem we actually have with NUnit 3 up through the 3.12 release. When the 3.13 release comes out, it will no longer be a problem, but of course the engine has to be able to deal with older releases as well.

As @rprouse points out, the proper place to detect and deal with problem 1 is in the driver. Only the driver has enough knowledge of the specific framework to be able to determine if the run has been terminated and (possibly) to take some action to terminate it. If it can't take action, it can return information so that it may be dealt with at a higher level, e.g. by killing a process.

Although I went ahead and added code to deal with problem 2 in the driver, it just didn't feel right. It seems to me that the second problem should be dealt with at a much higher level. If the run terminated but we didn't get notices for some of the active tests, then the engine needs to do something to generate those notices for runners that need them.

In fact, it turns out I already dealt with this problem in the engine build I am using for TestCentric. Here's my code from `MasterTestRunner`...

```c#
        /// <summary>
        /// Cancel the ongoing test run. If no  test is running, the call is ignored.
        /// </summary>
        /// <param name="force">If true, cancel any ongoing test threads, otherwise wait for them to complete.</param>
        public void StopRun(bool force)
        {
            _engineRunner.StopRun(force);

            // Frameworks should handle StopRun(true) by cancelling all tests and notifying
            // us of the completion of any tests that were running. However, this feature
            // may be absent in some frameworks or may be broken and we may not pass on the
            // notifications needed by some runners. In fact, such a bug is present in the
            // NUnit framework through release 3.12 and motivated the following code.
            //
            // We try to make up for the potential problem here by notifying the listeners 
            // of the completion of every pending WorkItem, one that started but never
            // sent a completion event. Since we have so far only noted this failure wrt
            // test suites and fixtures, those are the only ones we currently track.
            //
            // Note that this code only deals with notifications. If the framework did not
            // actually stop the run, that's a different problem, which has to be handled
            // at a lower level within the engine.

            if (force && !_workItemTracker.WaitForCompletion(WAIT_FOR_CANCEL_TO_COMPLETE))
            {
                _workItemTracker.IssuePendingNotifications(_eventDispatcher);

                // Indicate we are no longer running
                IsTestRunning = false;

                // Signal completion of the run
                _eventDispatcher.OnTestEvent($"<test-run id='{TestPackage.ID}' result='Failed' label='Cancelled' />");
            }
        }
```

The `WorkItemTracker` class tracks the start and completion of all test suites. It could do test cases too, but that has not been a problem so far. Its `IssuePendingNotifications` method creates completion events for any suites that have not yet sent them, so that the runner is happy. This takes care of our NUnit 3 issue as well as any problems that may arise with other frameworks all in one place.

One proviso... we are assuming, at this point, that the absence of notifications indicates we are dealing with problem 2. We have to do that because only the driver or lower-level runners can determine whether something has failed to terminate at all. For this PR, I believe it's reasonable to make that assumption and plan to deal with problem 1 separately.

In sum, I'd like to abandon the approach that handles this in the driver and port my existing code from the TestCentric build to upstream NUnit.

@nunit/engine-team Kindly comment. :smile:

**Comment by jnm2** [comment]
Sounds like tracking of outstanding items that may need a cancellation report is would be done at boundaries where communication with the framework may terminate in a non-cooperative fashion. Places that come to mind are if the engine uses Thread.Abort or otherwise abandons the framework, or if there is a process boundary w here the process may die which is actually running the tests. The old versions of the framework that didn't properly report cancellation could be considered for these purposes to also be terminating in a non-cooperative fashion.

This tracking could be applied just at each boundary (communication with old framework, communication with abortable framework, communication with separate process), but there are benefits to applying the tracking everywhere. We wouldn't have to do work to identify these conditions (e.g. old framework) at runtime. Potential future boundaries that we haven't thought of yet would already be handled, such as regressions in future versions of the framework. And doing things one way everywhere is simpler than having two options (track or don't track).

Would there be a way to make a change in one place that tracks and handles cancelling outstanding results in all these scenarios?

**Comment by CharliePoole** [comment]
@jnm2 Yes, there would be. I'm proposing to do it in the `MasterTestRunner` - and have actually done it already in the build I am using for the GUI. Because each report has to be parsed as XML in order to decide what to track it, there's a small performance hit to doing this.

OTOH, detecting success or failure in stopping the run would have to be done at the boundaries as you suggest. We'd probably have to inject a wait at that point.

An obvious point to catch a lot of things would be in `ProcessRunner`, since it could check to see if the process terminated and kill it otherwise. That would leave tests run in process as a problem.

**Comment by CharliePoole** [comment]
This discussion kind of died and I think it's a fairly important one. @ChrisMaddock @rprouse and @jnm2 already commented.

To kick it off again, I'm posting links to what I am currently doing for issue /TestCentric/testcentric-gui#374 I've tried to add comments that explain the motivation as well as the implementation of the change. While the issue is not yet resolved, I am now getting really good test output for hung tests, even though the framework fails to cancel them.

Here is the [WorkItemTracker class](https://github.com/TestCentric/testcentric-gui/blob/master/src/TestEngine/testcentric.engine/Runners/WorkItemTracker.cs), which tracks which tests have started and ended.

Here is the [code in `MasterTestRunner`](https://github.com/TestCentric/testcentric-gui/blob/issue-374/src/TestEngine/testcentric.engine/Runners/MasterTestRunner.cs#L179), which uses `WorkItemTracker`.

You'll notice that the current code unloads all engine runners, whether they need unloading or not. This does no harm but means they need to be reloaded the next time the user clicks run. A further iteration of this change would be to figure out __which__ assemblies need to be unloaded, but I think this is a good start.

__Question:__

Do we want to do something like this in the nunit engine. In the past, there has been some resistance because (I believe) folks want to trust the NUnit framework. But two things go against that logic: (1) we aim to run tests using multiple frameworks, some of which may not be fully developed and (2) this is a clear boundary between two projects and we should trust but try to verify. In this instance, we are only taking steps if we don't get a notice from the framework that the cancellation succeeded.

**Comment by CharliePoole** [comment]
@jnm2 One comment on your old comment... I don't think the engine ever aborts a framework thread. It merely tells the framework to abort. I could be wrong here, however, since I'm working with two different engines. I think I would say, however, that the engine __shouldn't__ abort threads created by the framework.

**Comment by rprouse** [comment]
The code that you linked to looks like a reasonable solution. It doesn't cross the line in that it doesn't cancel the framework threads, it just sends the event and stops waiting for them. My reluctance in the past has come from people wanting to kill all running test threads which seems dangerous. This does not do that. My only remaining concern is that we could end up with orphaned hung threads and with the VS adapter if it used this, orphaned VSTest processes. I'm not sure what we can do about that though and your solution is better than what we currently have.

**Comment by ChrisMaddock** [comment]
This seems sensible to me. 👍

**Comment by CharliePoole** [comment]
After some discussions we've had about the timing of the 4.0 release, I placed this into the 3.13 milestone. It's something I need for the GUI so I'll do a PR soon.
